검색 API에대해 TEST 코드 구현

KimJye opened the pull request:
## 작업 내용
검색 API에대해 TEST 코드를 구현했습니다.

## 실행 방법
npm test 

## 결과 첨부
![image](https://user-images.githubusercontent.com/33201628/70394399-2f2a1800-1a38-11ea-93a6-33d664050082.png)

KimJye committed:
[server] | feat(tagTest): tag test 코드 작성

KimJye committed:
[server] | feat(tag.service.test): tag service의 테스트코드 구현

tag service에서 findTags 함수에대한 테스트코드를 구현했습니다.
createQueryBuilder에대한 테스트가 잘 되지않아서, 기존에
createQueryBuilder로 구현되었던 서비스를 typeorm findAndCount로
변경하였습니다.

KimJye committed:
[server] | delete(tag.service.test): 불필요한 함수 삭제

KimJye committed:
[server] | delete(tag.service): 불필요한 함수 삭제

KimJye committed:
[server] | refactor(tag.service.test): 상대경로에서 절대경로로 변경

KimJye committed:
[server] | feat(tag.service.test): tag service 내용 추가로인한 코드 변경

status가 1인 값들만 보여줘야하기 때문에 status가 1인 조건을
추가하였습니다.
이에 따른 테스트코드도 변경되었습니다.

KimJye committed:
[server] | feat(tag.controller.test): 태그 컨트롤러 테스트

태그 컨트롤러 테스트를 작성했습니다.
하지만, 서비스에서 태그 엔티티 객체를 반환하여 테스트할때 어려움이
있습니다. 이 문제에대해서 같이 고민해봐야할것같습니다.

KimJye committed:
[server] | refactor(user.service): query 수정

createQueryBuilder로 테스트가 잘 되지않아서,
기존에 craeteQueryBuilder로 되어있던것을 typeorm의 메소드로
변경하였습니다.

KimJye committed:
[server] | feat(user.service.test): findUsers function test

keyword로 user를 찾는 함수를 테스트하였습니다.

KimJye committed:
[server] | refactor(video.service): typeorm 메소드로 수정

createQueryBuilder 테스트가 어려워서, createQueryBuilder로작성된 쿼리를
typeorm의 option들을 이용해서 수정했습니다.

KimJye committed:
[server] | feat(video.service.test): video service test 구현

video service에서 findVideos에대한 테스트코드를 구현했습니다.

KimJye committed:
[server] | refactor(tag.service.test): 상수 대문자로 변경

KimJye committed:
[server] | refactor(video.service.test): 불필요한 데이터 삭제

riverandeye commented on packages/server/src/user/user.service.ts:
```diff
@@ -25,36 +20,39 @@ export class UserService {
   ): Promise<[User[], number]> {
     const { page, keyword } = userListQueryDto;
 
     const offset = getOffset(page, USER_ITEMS_PER_PAGE);
```
```javascript
const offset = page ? getOffset(page, USER_ITEMS_PER_PAGE) : 0;
```
이렇게 지정하는건 어떨까요? page가 없는 경우엔 skip이 0인 것과 같으니, 밑의 함수에서 if문으로 나눌 필요가 없어질 것 같습니다!

riverandeye commented on packages/server/src/video/video.service.ts:
```diff
@@ -29,36 +26,25 @@ export class VideoService {
+          },
+          skip: offset,
+          take: VIDEO_ITEMS_PER_PAGE,
+        });
```
```javascript
    const relations = ['user'];
    const popularity = 'DESC';
    const id = 'DESC';
    const skip = page ? getOffset(page, VIDEO_ITEMS_PER_PAGE) : 0;
    const take = keyword || !page ? SEARCHED_ITEM_NUMBER : VIDEO_ITEMS_PER_PAGE;

    if (keyword) {
      return await this.videoRepository.findAndCount({
        relations,
        where: [
          { title: Like(`%${keyword}%`) },
          { description: Like(`%${keyword}%`), status: 1 },
        ],
        order: {
          popularity,
          id,
        },
        skip,
        take,
      });
    }

    const where = {
      status: 1,
    };

    if (sort === POPULAR && period !== PERIODS.all) {
      const startDatetime = moment()
        .subtract(...MOMENT_SUBTRACT_FROM_NOW_ARGUMENTS[period])
        .format(MOMENT_DATETIME_FORMAT);

      where['createdAt'] = MoreThan(startDatetime);
    }

    return await this.videoRepository.findAndCount({
      relations,
      where,
      order: {
        popularity,
        id,
      },
      skip,
      take,
    });
  }
```

해당 메소드가 return이 중복되고, 조건마다 달라지는 항목들을 쉽게 추적하기 위해서 이렇게 줄여봤는데, 검토 부탁드립니다.

KimJye replied:
반영했습니다 감사합니당

riverandeye reviewed (commented):
테스트 공부하시고 작성하시느라 고생 많으셨어요 ~.`

Atercatus commented on packages/server/src/tag/dto/tag-list-query.dto.ts:
```diff
@@ -1,6 +1,6 @@
 export class TagListQueryDto {
   public constructor(
-    public readonly page: number | string,
+    public readonly page?: number | string,
```
nullable 을 허용하지 않고 에러를 발생 시켜야 하지 않나요?

KimJye replied:
tag 검색경우 localhost:4000/tags?keyword=javascript 처럼 page가 query string에 없는 경우가 있습니다

Atercatus commented on packages/server/src/tag/tag.service.spec.ts:
```diff
@@ -0,0 +1,91 @@
+import { TagListQueryDto } from 'tag/dto/tag-list-query.dto';
+
+import { TAG_ITEMS_PER_PAGE, SEARCHED_ITEM_NUMBER } from 'common/constants';
+import { TAG_LIST, TAG_LIST_BY_KEYWORD_JAVA } from 'tag/tag.test.dummy.data';
```
import 부분은 외부 라이브러리는 상단에 고정하고 한줄 띄우고 내부구현한 모듈을 나열하면 좋을 거 같아요

Atercatus reviewed (changes requested):
테스트 관련 파일들은 `src` 보다는 루트에 존재하는 `test` 디렉토리에 있는 것이 좋을 거 같습니다. 소스코드와 테스트 코드를 분리하는게 관리하기에 더 용이하고 개발 과정에서 좀 더 편리함을 가져갈 수 있을 거 같아요.

hyokualexkwon reviewed (approved):
같이 논의했던, warning이 나타나지 않도록 describe의 로직을 test suite 마다 재 정의하는 방법을 적용해보면 좋을것 같습니다.

좋은 테스트네요! 감사합니다.

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/test_search_api_20191206

KimJye committed:
[server] | refactor(tag.teset): beforEach 상황별 추가

기존 코드는 findAndCount에대해 TAG_LIST를 반환하도록 beforEach에서
정의해두었습니다.
이에 문제점은 분기별로 다른 결과값이 나와야할때 주의가 떴었습니다.
그래서 이를 해결하고자 분기별로 beforEach를 각자 해주었습니다.

KimJye committed:
[server] | refactor(tag.test): import 정렬

KimJye committed:
[server] | refactor(service): 코드 리팩토링

기존에 if문으로 분기를 하였떤것을 삼항연산자로 변경하여 코드의 중복을
줄이고 가독성을 올렸습니다.
각 interface는 where 조건을 나타냅니다.

KimJye committed:
[server] | refactor(test): 서비스 리팩토링에의한 코드 수정

page가 없는 조건에서는 skip 옵션이 없었는데, 기존 코드를 삼항연산자로
변경하면서 skip 옵션이 추가되었습니다.

KimJye committed:
[server] | refactor(service, service.spec): 상대경로로 변경

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/test_search_api_20191206

KimJye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/test_search_api_20191206
